src/fastpath/fastpath.c

```diff
@@ -502,21 +696,45 @@ void NORETURN fastpath_reply_recv(word_t cptr, word_t msgInfo)
     callerSlot->cteMDBNode = nullMDBNode;
 #endif
 
-    /* I know there's no fault, so straight to the transfer. */
+#ifdef CONFIG_EXCEPTION_FASTPATH
+    if (unlikely(fault_type != seL4_Fault_NullFault)) {
+        /* Note - this works as is for VM faults but will need to be changed when other faults are added. VM faults always
+         * restart the faulting thread upon reply but this is not always the case with other types of faults. This can either
+         * be handled in the fastpath or redirected to the slowpath, but either way, this code must be changed so we do not
+         * forcefully switch to a thread which is meant to stay inactive. */
 
-    /* Replies don't have a badge. */
-    badge = 0;
+        word_t pc = getRestartPC(caller);
+        setNextPC(caller, pc);
 
-    fastpath_copy_mrs(length, NODE_STATE(ksCurThread), caller);
+        /* Clear the tcbFault variable to indicate that it has been handled. */
+        caller->tcbFault = seL4_Fault_NullFault_new();
 
-    /* Dest thread is set Running, but not queued. */
-    thread_state_ptr_set_tsType_np(&caller->tcbState,
-                                   ThreadState_Running);
-    switchToThread_fp(caller, cap_pd, stored_hw_asid);
+        /* Dest thread is set Running, but not queued. */
```

Comment:
Can you expand the comment here to mention that VMFaults set thread's to ThreadState_Restart in the slowpath, but this is inlining the part of activateThread() that moves a thread from Restart to Running while updating the PC to RestartPC?